packages/circuits/src/initial.circom

```diff
@@ -0,0 +1,89 @@
+pragma circom 2.0.0;
+
+include "./merkle.circom";
+include "../node_modules/circomlib/circuits/comparators.circom";
+
+template CheckReplaceArray(n) {
+    signal input index;
+    signal input newValue;
+    signal input array[n];
+    signal output newArray[n];
+    signal output replacedValue;
+
+    component checkIndex[n];
+    signal items[n];
+    for (var i = 0; i < n; i++) {
+        checkIndex[i] = IsEqual();
+        checkIndex[i].in[0] <== i;
+        checkIndex[i].in[1] <== index;
+        newArray[i] <== (newValue - array[i]) * checkIndex[i].out + array[i];
+        // newArray[i] <== (checkIndex[i].out * newValue) + ((1 - checkIndex[i].out) * array[i]);
+        items[i] <== checkIndex[i].out * array[i];
+    }
+
+    // get the replaced value
+    var temp = 0;
+    for (var i = 0; i < n; i++) {
+        temp += items[i];
+    }
+    replacedValue <== temp;
+}
+
+template Initial(levels, cardCount) {
+    signal input deckRoot;
+    signal input newDeckRoot;
+    signal input deckLeaves[2**levels];
+    signal input initialDeckTailCardIndex;
+    signal input handRoot;
+    signal input newHandRoot;
+    signal input drawnCardIndices[cardCount];
+
+    component checkInitialDeck = CheckMerkleRoot(levels);
+    checkInitialDeck.root <== deckRoot;
+    checkInitialDeck.leaves <== deckLeaves;
+
+    // check initial hand is all null
+    var handLeaves[2**levels];
+    for (var i = 0; i < 2**levels; i++) {
+        handLeaves[i] = 255;
+    }
+    component checkInitialHand = CheckMerkleRoot(levels);
+    checkInitialHand.root <== handRoot;
+    checkInitialHand.leaves <== handLeaves;
+
+    component updateDeckArray[2*cardCount];
+    component updateHandArray[cardCount];
+    var tailCardIndex = initialDeckTailCardIndex;
+    for (var i = 0; i < 2*cardCount; i+=2) {
+        // replace tail card with null
+        updateDeckArray[i] = CheckReplaceArray(2**levels);
+        updateDeckArray[i].index <== tailCardIndex;
+        updateDeckArray[i].newValue <== 255;
+        updateDeckArray[i].array <== i == 0 ? deckLeaves : updateDeckArray[i-1].newArray;
+
+        // replace the drawn card with tail card
+        updateDeckArray[i+1] = CheckReplaceArray(2**levels);
+        updateDeckArray[i+1].index <== drawnCardIndices[i/2];
+        updateDeckArray[i+1].newValue <== updateDeckArray[i].replacedValue;
+        updateDeckArray[i+1].array <== updateDeckArray[i].newArray;
+
+        // update tail card index
+        tailCardIndex--;
+
+        // update hand with the drawn card
+        updateHandArray[i/2] = CheckReplaceArray(2**levels);
+        updateHandArray[i/2].index <== i/2;
+        updateHandArray[i/2].newValue <== updateDeckArray[i+1].replacedValue;
+        updateHandArray[i/2].array <== i == 0 ? handLeaves : updateHandArray[i/2-1].newArray;
```

Comment:
Couldn't we get rid of all the iterations over the deck & hand? Currently this iterates over all drawn cards, and then for every one of them generates two iterations over the whole deck and one over the hand (via three `CheckReplaceArray` components) to get the updated deck & hand.

What if instead we still iterate over the cards drawn, but directly set the values in an `updatedDeck` and an `updatedHand` array variables?

Reply:
That was my first approach, but apparently an annoying thing about Circom is that all arrays cannot use values that are unknown during compilation, which will lead to non quadratic constraint error. So for example if you ever assign something like `updateDeck[index] = x`, it will become non quadratic. 

I do feel that my approach is actually pretty inefficient hahaha, so keen to get maybe @Boffee thoughts on this. (Kudos to him for pointing out about the array index issue, I was stuck on that for a long time TT)

Reply:
I spent some time thinking on it, I feel like a better approach here might be to iterate the draw proof as many times as there are cards to draw.

Each draw proof basically needs to do a Merkle proof for the card to remove + for the last card (`tailCard` in the implem), as well as construct two Merkle roots (one for copying the tail card to the removed card index, one to remove the tail card). Each of these operation have log complexity in the size of the deck, so you pay `O(4 * initialHandSize * log(deckSize))` where we can constrain `log(deckSize)`to 6 in practice (assuming max 64 cards per deck). So for a hand of 7 cards and a deck of 60 cards, `168c` where `c` is some constant number of ops from the implem of membership checks and root construction.

In contrast, the current approach, does `O(2 * deckSize * initialHandSize)` so `840c` (of course one also need to check to see if the overhead is different ...).

A naive approach would need Merkle branch for every removed card + tailCard to be provided, but in reality we can recompute the whole Merkle tree ourselves + compute the Merkle branch dynamically (zero constraints for Merkle branch computation, using `<--` because everything is ultimately authenticated against the Merkle roots, and each successive Merkle root is correct by construction).

By the way, I looked at the code for the draw proof, and there are quite some problems there also — for one, I found a pretty bad bug I think (the code does not actually computes the second Merkle root as it should, instead I just does a membership check against it). It also does not ensure that the `tailCard` that is passed is actually the last card in the deck (this could be fine if we say that the remaining deck size is tracked in the contract and use that to supply the last card index to the verifier).

Reply:
As is often the case, I had to post this to realize how stupid it was: the overhead from the hashing in the Merkle root probably makes the 4-5x gain obsolete. Oh well.

Reply:
Just did a comparison for using Merkle root as temporary state instead of using a deck array. The constraints of the circuits are as follows: 
1. Deck array update (current approach): ~180k constraints
2. Merkle root update : ~390k constraints

From what I understand, number of constraints is an indication of optimization of circuit, so perhaps the merkle approach is more expensive. 180k is still a large circuit though, for comparison our current other circuits draw.circom is at ~55k and play.circom is at ~31k